app/searchers/application_searcher.rb

```diff
@@ -0,0 +1,113 @@
+class ApplicationSearcher
+  include Enumerable
+
+  class << self
+    attr_accessor :parser_klass, :index_klass, :fields, :keywords, :field_type_mappings
+
+    def index klass
+      @index_klass = klass
+    end
+
+    def keyword key, aliases: nil, as: nil
+      keywords_data << { key => { aliases: aliases, as: as } }
+    end
+
+    def keywords_data
+      @keywords_data ||= []
+    end
+  end
+
+  attr_reader :input
+  attr_reader :errors
+
+  def initialize
+    @errors = Errors.new
+  end
+
+  def query str
+    @input = str
+    self
+  end
+
+  def scope raw_es_query
+    @scope = raw_es_query
+    self
+  end
+
+  def results
+    raw_results = index_klass.query build_elasticsearch_query.to_elasticsearch
+    raw_results.query @scope
+    fetch raw_results
+  end
+
+  def each
+    results.each
+  end
+
+  protected
+
+  def index_klass
+    self.class.index_klass
+  end
+
+  def keywords_data
+    self.class.keywords_data
```

Comment:
this feels like a bunch of misdirection

Reply:
the name or the class attribute accessors inside the instance?